imap_processing/ialirt/l0/process_codicehi.py

```diff
@@ -0,0 +1,156 @@
+    #   * group    (group) int64 7kB 1 1 1 1 1 1 1 1 1 ... 15 15 15 15 15 15 15 15 15
+    grouped_data["group"] = ("group", group_labels)
+
+    return grouped_data
+
+
+def append_cod_hi_data(dataset: xr.Dataset) -> xr.Dataset:
+    """
+    Append the cod_hi_## data values and create an xarray.
+
+    Parameters
+    ----------
+    dataset : xr.Dataset
+        Original dataset of group.
+
+    Returns
+    -------
+    appended_dataset : xr.Dataset
+        Dataset with cod_hi_## stacked.
+    """
+    # Number of codice hi data rows
+    num_cod_hi_rows = 5
+    cod_hi_data = np.stack(
+        [dataset[f"cod_hi_data_{i:02}"].values for i in range(num_cod_hi_rows)], axis=1
+    )
+
+    repeated_data = {
+        var: np.repeat(dataset[var].values, num_cod_hi_rows)
+        for var in dataset.data_vars
+        if not var.startswith("cod_hi_data_")
+    }
+
+    repeated_data["cod_hi_appended"] = cod_hi_data.flatten()
+    repeated_epoch = np.repeat(dataset["epoch"].values, num_cod_hi_rows)
+
+    appended_dataset = xr.Dataset(
+        data_vars={name: ("epoch", values) for name, values in repeated_data.items()},
+        coords={"epoch": repeated_epoch},
+    )
+
+    return appended_dataset
+
+
+def process_codicehi(xarray_data: xr.Dataset) -> list[dict]:
+    """
+    Create final data products.
+
+    Parameters
+    ----------
+    xarray_data : xr.Dataset
+        Parsed data.
+
+    Returns
+    -------
+    codicehi_data : list[dict]
+        Dictionary of final data product.
+
+    Notes
+    -----
+    This function is incomplete and will need to be updated to include the
+    necessary calculations and data products.
+    - Calculate species counts (pg 27 of Algorithm Document)
+    - Calculate rates (assume 4 minutes per group)
+    - Calculate L2 CoDICE pseudodensities (pg 37 of Algorithm Document)
+    - Calculate the public data products
+    """
+    grouped_data = find_groups(xarray_data)
+    unique_groups = np.unique(grouped_data["group"])
+    codicehi_data: list[dict[str, Any]] = [{}]
+
+    for group in unique_groups:
+        # cod_hi_counter values for the group should be 0-233 with no duplicates.
+        subcom_values = grouped_data["cod_hi_counter"][
+            (grouped_data["group"] == group).values
+        ]
+
+        # Ensure no duplicates and all values from 0 to 233 are present
+        if not np.array_equal(subcom_values, np.arange(234)):
+            logger.warning(
+                f"Group {group} does not contain all values from 0 to "
+                f"233 without duplicates."
+            )
+            continue
+
+        mask = grouped_data["group"] == group
+        filtered_indices = np.where(mask)[0]
+        group_data = grouped_data.isel(epoch=filtered_indices)
+
+        append_cod_hi_data(group_data)
+
+        # TODO: calculate species counts
+        # TODO: calculate rates
+        # TODO: calculate L2 CoDICE pseudodensities
+        # TODO: calculate the public data products
```

Comment:
Thanks for putting these here, it makes it clear where the functions in my pipeline will eventually fit in!